routing/endpointregistry.go

```diff
@@ -82,58 +81,43 @@ func NewEndpointRegistry(o RegistryOptions) *EndpointRegistry {
 		o.LastSeenTimeout = defaultLastSeenTimeout
 	}
 
-	return &EndpointRegistry{
-		data:            map[string]*entry{},
-		lastSeen:        map[string]time.Time{},
+	result := &EndpointRegistry{
+		data:            sync.Map{},
 		lastSeenTimeout: o.LastSeenTimeout,
 		now:             time.Now,
 	}
+
+	for host, detected := range o.Detected {
+		e := &entry{detected: detected, lastSeen: result.now()}
+		result.data.Store(host, e)
+	}
+
+	return result
 }
 
 func (r *EndpointRegistry) GetMetrics(key string) Metrics {
-	r.mu.Lock()
-	defer r.mu.Unlock()
-
-	e := r.getOrInitEntryLocked(key)
+	e, _ := r.data.LoadOrStore(key, &entry{})
 	copy := &entry{}
-	*copy = *e
+	*copy = *e.(*entry)
```

Comment:
This is not thread-safe. I think we should create a test that calls GetMetrics and IncInflightRequest in parallel and it should not fail under race detector.

Reply:
https://github.com/zalando/skipper/pull/2759/files#diff-92e82ef5e7102bf8a84e0f19a069277acf1b50b758a92fd35d62521d45378b2aR96